local-cluster/tests/local_cluster.rs

```diff
@@ -1680,7 +1601,282 @@ fn test_validator_saves_tower() {
+        sleep(Duration::from_millis(100));
+        let highest_bank = client
+            .get_slot_with_commitment(CommitmentConfig::recent())
+            .unwrap();
+        if highest_bank >= next_slot_on_a {
+            break;
+        }
+    }
+    let validator_a_info = cluster.exit_node(&validator_a_pubkey);
+    let client = cluster.get_validator_client(&validator_b_pubkey).unwrap();
+    loop {
+        sleep(Duration::from_millis(100));
+        let highest_bank = client
+            .get_slot_with_commitment(CommitmentConfig::recent())
+            .unwrap();
+        if highest_bank >= next_slot_on_a {
+            break;
+        }
+    }
+    let _validator_c_info = cluster.exit_node(&validator_b_pubkey);
+
+    // Step 2:
+    // Stop validator and truncate ledger
+    info!("truncate validator C's ledger");
+    {
+        // first copy from validator A's ledger
+        std::fs::remove_dir_all(&validator_c_info.info.ledger_path).unwrap();
+        let mut opt = fs_extra::dir::CopyOptions::new();
+        opt.copy_inside = true;
+        fs_extra::dir::copy(
+            &validator_a_info.info.ledger_path,
+            &validator_c_info.info.ledger_path,
+            &opt,
+        )
+        .unwrap();
+        std::fs::remove_file(Tower::get_filename(
+            &validator_c_info.info.ledger_path,
+            &validator_a_pubkey,
+        ))
+        .unwrap();
+
+        let blockstore = Blockstore::open_with_access_type(
+            &validator_c_info.info.ledger_path,
+            AccessType::PrimaryOnly,
+            None,
+        )
+        .unwrap_or_else(|e| {
+            panic!(
+                "Failed to open ledger at {:?}, err: {}",
+                validator_c_info.info.ledger_path, e
+            );
+        });
+        blockstore.purge_from_next_slots(base_slot + 1, base_slot + truncated_slots);
+        blockstore.purge_slots(base_slot + 1, base_slot + truncated_slots, PurgeType::Exact);
+    }
+    info!("truncate validator A's ledger");
+    {
+        let blockstore = Blockstore::open_with_access_type(
+            &validator_a_info.info.ledger_path,
+            AccessType::PrimaryOnly,
+            None,
+        )
+        .unwrap_or_else(|e| {
+            panic!(
+                "Failed to open ledger at {:?}, err: {}",
+                validator_a_info.info.ledger_path, e
+            );
+        });
+        blockstore.purge_from_next_slots(next_slot_on_a + 1, next_slot_on_a + truncated_slots);
+        blockstore.purge_slots(
+            next_slot_on_a + 1,
+            next_slot_on_a + truncated_slots,
+            PurgeType::Exact,
+        );
+        if !with_tower {
+            info!("Removing tower!");
+            std::fs::remove_file(Tower::get_filename(
+                &validator_a_info.info.ledger_path,
+                &validator_a_pubkey,
+            ))
+            .unwrap();
+
+            // For some reason, fork_choice always selects slot 27 over votes_on_c_fork.
```

Comment:
@ryoqun I think I see what's happening.

 If you don't delete slot 27 from the ledger, then the validator `A` will immediately vote for 27 on restart, because it hasn't gotten the heavier fork from validator `C` yet. Then it will be stuck on `27` unable to switch because `C` doesn't have enough stake to generate a switching proof

Reply:
cool. that makes sense. The hard part of writing this was that validator actually try very hard to move forward the chain and resolve any divergent. I guess I can't hold validator from voting after restart, right? I'll update this comment of `for some reason` with your comment. :)

Reply:
done!

Reply:
@ryoqun yeah no way to prevent the validator from voting on restart, so I think this is the correct thing!